lib/ts/controllers/s-uploader.ts

```diff
@@ -0,0 +1,174 @@
+namespace Stacks {
+    interface FilePreview {
+        data?: string | ArrayBuffer;
+        name: string;
+        type: string;
+    };
+
+    export class UploaderController extends Stacks.StacksController {
+        static targets = ["input", "previews", "uploader"];
+        private inputTarget!: HTMLInputElement;
+        private previewsTarget!: HTMLElement;
+        private uploaderTarget!: HTMLElement;
+
+        private boundDragEnter!: any;
+        private boundDragLeave!: any;
+
+        private static readonly FILE_DISPLAY_LIMIT = 10;
+        private static readonly MAX_FILE_SIZE = 1024 * 1024 * 10; // 10 MB
+
+        connect() {
+            super.connect();
+            this.boundDragEnter = this.handleUploaderActive.bind(this, true);
+            this.boundDragLeave = this.handleUploaderActive.bind(this, false);
+
+            this.inputTarget.addEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.addEventListener("dragleave", this.boundDragLeave);
+        }
+
+        disconnect() {
+            this.inputTarget.removeEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.removeEventListener("dragleave", this.boundDragLeave);
+            super.disconnect();
+        }
+
+        /**
+         * Handles rendering the file preview state on input change
+         */
+        handleInput() {
+            this.previewsTarget.innerHTML = "";
+
+            if (!this.inputTarget.files) {
+                return;
+            }
+
+            const count = this.inputTarget.files.length;
+                this.getDataURLs(this.inputTarget.files, UploaderController.FILE_DISPLAY_LIMIT)
+                    .then((res) => {
```

Comment:
Thanks for killing the `any` I plopped in there @b-kelly. Would this be reasonable?

```suggestion
                    .then((res: FilePreview[]) => {
```

Reply:
Yes, absolutely. In this case (for me, locally anyways) TS was able to infer the type, so I dropped it. You're welcome to add it back according to your preferences - I don't have a strong opinion on leaving it in or out here.